### ISSUE TITLE
tft unable to set target to a list of strings (multiple targets)

### ISSUE THREAD

**Issue by QitianMa** [jdb78/pytorch-forecasting#542]
- PyTorch-Forecasting version: 0.8.5
- PyTorch version: 1.8.1
- Python version: 3.8.10
- Operating System: linux 3.10.0-1160.25.1.el7.x86_64

according to doc: https://pytorch-forecasting.readthedocs.io/en/latest/api/pytorch_forecasting.data.timeseries.TimeSeriesDataSet.html
the target parameter can be set to a list of strings indicating multiple variables for prediction
However, during run time, the code spits the following error:

```
TypeError: new() received an invalid combination of arguments - got (list, int), but expected one of:
 * (*, torch.device device)
      didn't match because some of the arguments have invalid types: (list, int)
 * (torch.Storage storage)
 * (Tensor other)
 * (tuple of ints size, *, torch.device device)
 * (object data, *, torch.device device)
```

**Comment by QitianMa** [comment]
TypeError                                 Traceback (most recent call last)
~/HI/Syrto/tft_training.py in 
    195 logger = TensorBoardLogger(save_dir=params["save"]["output_dir"], name=params["save"]["trial_version"])
    196 
--> 197 tft = TemporalFusionTransformer.from_dataset(
    198     train_set,
    199     learning_rate=params["train"]["learning_rate"],

~/miniconda3/lib/python3.9/site-packages/pytorch_forecasting/models/temporal_fusion_transformer/__init__.py in from_dataset(cls, dataset, allowed_encoder_known_variable_names, **kwargs)
    350 
    351         # create class and return
--> 352         return super().from_dataset(
    353             dataset, allowed_encoder_known_variable_names=allowed_encoder_known_variable_names, **new_kwargs
    354         )

~/miniconda3/lib/python3.9/site-packages/pytorch_forecasting/models/base_model.py in from_dataset(cls, dataset, allowed_encoder_known_variable_names, **kwargs)
   1295         )
   1296         new_kwargs.update(kwargs)
-> 1297         return super().from_dataset(dataset, **new_kwargs)
   1298 
   1299     def calculate_prediction_actual_by_variable(

~/miniconda3/lib/python3.9/site-packages/pytorch_forecasting/models/base_model.py in from_dataset(cls, dataset, **kwargs)
    847         if "output_transformer" not in kwargs:
    848             kwargs["output_transformer"] = dataset.target_normalizer
--> 849         net = cls(**kwargs)
    850         net.dataset_parameters = dataset.get_parameters()
    851         if dataset.multi_target:

~/miniconda3/lib/python3.9/site-packages/pytorch_forecasting/models/temporal_fusion_transformer/__init__.py in __init__(self, hidden_size, lstm_layers, dropout, output_size, loss, attention_head_size, max_encoder_length, static_categoricals, static_reals, time_varying_categoricals_encoder, time_varying_categoricals_decoder, categorical_groups, time_varying_reals_encoder, time_varying_reals_decoder, x_reals, x_categoricals, hidden_continuous_size, hidden_continuous_sizes, embedding_sizes, embedding_paddings, embedding_labels, learning_rate, log_interval, log_val_interval, log_gradient_flow, reduce_on_plateau_patience, monotone_constaints, share_single_variable_networks, logging_metrics, **kwargs)
    322             )
    323         else:
--> 324             self.output_layer = nn.Linear(self.hparams.hidden_size, self.hparams.output_size)
    325 
    326     @classmethod

~/miniconda3/lib/python3.9/site-packages/torch/nn/modules/linear.py in __init__(self, in_features, out_features, bias)
     77         self.in_features = in_features
     78         self.out_features = out_features
---> 79         self.weight = Parameter(torch.Tensor(out_features, in_features))
     80         if bias:
     81             self.bias = Parameter(torch.Tensor(out_features))

TypeError: new() received an invalid combination of arguments - got (list, int), but expected one of:
 * (*, torch.device device)
      didn't match because some of the arguments have invalid types: (!list!, !int!)
 * (torch.Storage storage)
 * (Tensor other)
 * (tuple of ints size, *, torch.device device)
 * (object data, *, torch.device device)

**Comment by jdb78** [comment]
Could you post the full code of the TimeSeriesDataSet definition?

**Comment by QitianMa** [comment]
> Could you post the full code of the TimeSeriesDataSet definition?

https://github.com/jdb78/pytorch-forecasting/blob/master/pytorch_forecasting/data/timeseries.py

**Comment by QitianMa** [comment]
> Could you post the full code of the TimeSeriesDataSet definition?

train_set = TimeSeriesDataSet(
    df_train[lambda x: x.bilancio_year <= params["train"]["training_cutoff"]],
    group_ids=params["tft"]["group_ids"],
    target=params["tft"]["target"],
    time_idx=params["tft"]["time_idx"],
    min_encoder_length=1,
    max_encoder_length=10,
    min_prediction_length=1,
    max_prediction_length=1,
    categorical_encoders={},
    static_categoricals=params["features"]["static_categoricals"],
    time_varying_known_categoricals=params["features"]["time_varying_known_categoricals"],
    time_varying_known_reals=params["features"]["time_varying_known_reals"],
#     time_varying_unknown_categoricals=["status","listing"],
    time_varying_unknown_categoricals=params["features"]["time_varying_unknown_categoricals"],
    time_varying_unknown_reals=params["features"]["time_varying_unknown_reals"]
)

**Comment by QitianMa** [comment]
target is a list of variable names such as ["a", "b", "c"]

**Comment by jdb78** [comment]
What do you get for `train_set.target_names`? I feel something is going wrong also with the loss. When you creating the tft, are you using explicitly a loss function?

**Comment by QitianMa** [comment]
> What do you get for `train_set.target_names`? I feel something is going wrong also with the loss. When you creating the tft, are you using explicitly a loss function?

train_set.target_names: ['Turnover', 'FixedAssets']
loss is QuantileLoss()

```
ft = TemporalFusionTransformer.from_dataset(
    train_set,
    learning_rate=params["train"]["learning_rate"],
    hidden_size=params["tft"]["hidden_size"],
    lstm_layers=params["tft"]["lstm_layers"],
    attention_head_size=params["tft"]["heads"],
    dropout=params["train"]["dropout"],
    hidden_continuous_size=8,
    optimizer = params["train"]["optimizer"],
    # output_size=7
    loss=QuantileLoss(),
    log_interval=10,
    log_val_interval=1,
    # reduce_on_plateau_patience=3,
)
```

**Comment by jdb78** [comment]
Hm. Something goes wrong with the loss definition. Think I found the issue. Could you confirm that it runs without an issue if you try define the loss as
`loss=MultiLoss([QuantileLoss(), QuantileLoss()])`

**Comment by QitianMa** [comment]
> Hm. Something goes wrong with the loss definition. Think I found the issue. Could you confirm that it runs without an issue if you try define the loss as
> `loss=MultiLoss([QuantileLoss(), QuantileLoss()])`

Hi I have got a new error as follows. Such error does not exist for single target.
```
Traceback (most recent call last):
  File "tft_training.py", line 235, in <module>
    val_dataloaders=testin_loader,
  File "/home/3068020/.local/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py", line 458, in fit
    self._run(model)
  File "/home/3068020/.local/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py", line 756, in _run
    self.dispatch()
  File "/home/3068020/.local/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py", line 797, in dispatch
    self.accelerator.start_training(self)
  File "/home/3068020/.local/lib/python3.7/site-packages/pytorch_lightning/accelerators/accelerator.py", line 96, in start_training
    self.training_type_plugin.start_training(trainer)
  File "/home/3068020/.local/lib/python3.7/site-packages/pytorch_lightning/plugins/training_type/training_type_plugin.py", line 144, in start_training
    self._results = trainer.run_stage()
  File "/home/3068020/.local/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py", line 807, in run_stage
    return self.run_train()
  File "/home/3068020/.local/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py", line 869, in run_train
    self.train_loop.run_training_epoch()
  File "/home/3068020/.local/lib/python3.7/site-packages/pytorch_lightning/trainer/training_loop.py", line 489, in run_training_epoch
    batch_output = self.run_training_batch(batch, batch_idx, dataloader_idx)
  File "/home/3068020/.local/lib/python3.7/site-packages/pytorch_lightning/trainer/training_loop.py", line 728, in run_training_batch
    self.optimizer_step(optimizer, opt_idx, batch_idx, train_step_and_backward_closure)
  File "/home/3068020/.local/lib/python3.7/site-packages/pytorch_lightning/trainer/training_loop.py", line 432, in optimizer_step
    using_lbfgs=is_lbfgs,
  File "/home/3068020/.local/lib/python3.7/site-packages/pytorch_lightning/core/lightning.py", line 1403, in optimizer_step
    optimizer.step(closure=optimizer_closure)
  File "/home/3068020/.local/lib/python3.7/site-packages/pytorch_lightning/core/optimizer.py", line 214, in step
    self.__optimizer_step(*args, closure=closure, profiler_name=profiler_name, **kwargs)
  File "/home/3068020/.local/lib/python3.7/site-packages/pytorch_lightning/core/optimizer.py", line 134, in __optimizer_step
    trainer.accelerator.optimizer_step(optimizer, self._optimizer_idx, lambda_closure=closure, **kwargs)
  File "/home/3068020/.local/lib/python3.7/site-packages/pytorch_lightning/accelerators/accelerator.py", line 329, in optimizer_step
    self.run_optimizer_step(optimizer, opt_idx, lambda_closure, **kwargs)
  File "/home/3068020/.local/lib/python3.7/site-packages/pytorch_lightning/accelerators/accelerator.py", line 336, in run_optimizer_step
    self.training_type_plugin.optimizer_step(optimizer, lambda_closure=lambda_closure, **kwargs)
  File "/home/3068020/.local/lib/python3.7/site-packages/pytorch_lightning/plugins/training_type/training_type_plugin.py", line 193, in optimizer_step
    optimizer.step(closure=lambda_closure, **kwargs)
  File "/home/3068020/.local/lib/python3.7/site-packages/torch/optim/optimizer.py", line 89, in wrapper
    return func(*args, **kwargs)
  File "/home/3068020/.local/lib/python3.7/site-packages/torch/autograd/grad_mode.py", line 27, in decorate_context
    return func(*args, **kwargs)
  File "/home/3068020/.local/lib/python3.7/site-packages/torch/optim/adam.py", line 66, in step
    loss = closure()
  File "/home/3068020/.local/lib/python3.7/site-packages/pytorch_lightning/trainer/training_loop.py", line 723, in train_step_and_backward_closure
    split_batch, batch_idx, opt_idx, optimizer, self.trainer.hiddens
  File "/home/3068020/.local/lib/python3.7/site-packages/pytorch_lightning/trainer/training_loop.py", line 826, in training_step_and_backward
    self.backward(result, optimizer, opt_idx)
  File "/home/3068020/.local/lib/python3.7/site-packages/pytorch_lightning/trainer/training_loop.py", line 860, in backward
    result.closure_loss, optimizer, opt_idx, should_accumulate, *args, **kwargs
  File "/home/3068020/.local/lib/python3.7/site-packages/pytorch_lightning/accelerators/accelerator.py", line 309, in backward
    self.lightning_module, closure_loss, optimizer, optimizer_idx, should_accumulate, *args, **kwargs
  File "/home/3068020/.local/lib/python3.7/site-packages/pytorch_lightning/plugins/precision/precision_plugin.py", line 79, in backward
    model.backward(closure_loss, optimizer, opt_idx)
  File "/home/3068020/.local/lib/python3.7/site-packages/pytorch_lightning/core/lightning.py", line 1275, in backward
    loss.backward(*args, **kwargs)
  File "/home/3068020/.local/lib/python3.7/site-packages/torch/tensor.py", line 245, in backward
    torch.autograd.backward(self, gradient, retain_graph, create_graph, inputs=inputs)
  File "/home/3068020/.local/lib/python3.7/site-packages/torch/autograd/__init__.py", line 147, in backward
    allow_unreachable=True, accumulate_grad=True)  # allow_unreachable flag
RuntimeError: element 0 of tensors does not require grad and does not have a grad_fn
```

**Comment by jdb78** [comment]
Guess this does not happen immediately, right? It is most likely divergence leading to a nan without a grad. So probably a too high learning rate or you need to clip your gradients at a lower value.

**Comment by QitianMa** [comment]
> Guess this does not happen immediately, right? It is most likely divergence leading to a nan without a grad. So probably a too high learning rate or you need to clip your gradients at a lower value.

This happens immediately. I tried gradient_clip_val=0.0001, and learning rate 1e-6. This issue persists.

**Comment by TimothyLiuu** [comment]
Hello! Have you solve the problem? @QitianMa

**Comment by theoctopusride** [comment]
Any success with this? @LumingSun @QitianMa @jdb78 I am running into a separate issue described here: https://github.com/jdb78/pytorch-forecasting/issues/820#issuecomment-1137898515

**Comment by AdolHong** [comment]
FYI,   I met error 'element 0 of tensors does not require grad and does not have a grad_fn',  when I using encoderNormalizer while setting min_encoder_length = 1.

**Comment by theoctopusride** [comment]
@AdolHong what happens if you do not set the target normalizer? According to the docs for Time series data set [here](https://pytorch-forecasting.readthedocs.io/en/stable/data.html#time-series-data-set):

By default an appropriate normalizer is chosen automatically.

**Comment by AIME2021KE** [comment]
I'm brand new to using TFT and I'm having trouble using it with multiple targets (columns of data). I'm working from the Stallion Demo / Data (https://pytorch-forecasting.readthedocs.io/en/latest/tutorials/stallion.html) but increasing the target to a list of targets. I've successfully created the TimeSeriesDataSet, but when I try to initialize the TemporalFusionTransformer.from_dataset it crashes. I'm sure it's about my improper initialization, but lost the thread at the __init__ of  the MultiEmbedding and hparams. Is there an end to end demo of using TFT with multiple targets like the Stallion demo that shows what is required for this case? I've provided the stripped (from original demo try) at 
https://github.com/AIME2021KE/TFT_multitarget_try
as well as the error message I get. Can provide both as .txt files here if that is preferred.
